### PR TITLE
Support derivation for groupBy online fetch

### PR DESCRIPTION
**Pull request by hanyuli1995**
## Summary

Currently feature derivation is only supported for online join. We want to support derivation for online groupBy fetch as well. 

After the code change, users will be able to 
1. Fetch group by data with defined derivation rules.
2. Fetch join data with defined derivation rules in join part.

## Test Plan

- [x] Added Unit Tests
- [x] Covered by existing CI
- [x] Integration tested
For Airbnb useres, the local tests result are recorded in 
https://docs.google.com/document/d/1uuMkJ8WpQy3XPgfu89OsgdloLdwtLRhNSN6FE2XsHCY/edit?usp=sharing

## Checklist
- [x] Documentation update

## Reviewers
@better365 @nikhilsimha

### CODE REVIEW

File: online/src/main/scala/ai/chronon/online/FetcherBase.scala [hunk 1]
```diff
@@ -309,14 +305,85 @@ class FetcherBase(kvStore: KVStore,
                   ex.printStackTrace()
                   throw ex
               }
+              logger.info(s"[test] groupByServingInfo ${groupByServingInfo}")
```
Comment (nikhilsimha) [inline online/src/main/scala/ai/chronon/online/FetcherBase.scala:308]: assuming that these log statements will be removed before merging

File: online/src/main/scala/ai/chronon/online/FetcherBase.scala [hunk 2]
```diff
@@ -309,14 +305,85 @@ class FetcherBase(kvStore: KVStore,
...
+    val requestDs = TsUtils.toStr(requestTs).substring(0, 10)
+    // used for derivation based on ts/ds
+    val tsDsMap: Map[String, AnyRef] =
+      Map("ts" -> (requestTs).asInstanceOf[AnyRef], "ds" -> (requestDs).asInstanceOf[AnyRef])
+    val timeFields: Array[StructField] = Array(
+      StructField("ts", LongType),
+      StructField("ds", StringType)
+    )
+    val conf = groupByServingInfo.groupBy
+    def deriveFunc: (Map[String, Any], Map[String, Any]) => Map[String, Any] = {
```
Comment (nikhilsimha) [inline online/src/main/scala/ai/chronon/online/FetcherBase.scala:356]: running this function is expensive - we we need to cache the result of the derive function. You can follow how it is done in the join derivation case.

Comment (nikhilsimha) [reply online/src/main/scala/ai/chronon/online/FetcherBase.scala:356]: I notice that this code is repeated - feel free to refactor to re-use both for join derivations and groupBy derivations.

File: online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    val exceptions: Map[String, Any] = preDerivation.iterator.filter(_._1.endsWith("_exception")).toMap
+    if (exceptions.isEmpty) {
+      return derived
+    }
+    val exceptionParts: Array[String] = exceptions.keys.map(_.dropRight("_exception".length)).toArray
+    derived.filterKeys(key => !exceptionParts.exists(key.startsWith)).toMap ++ exceptions
+  }
+
+  def buildRenameOnlyDerivationFunction(derivationsScala: List[Derivation]): DerivationFunc = {
+    {
```
Comment (nikhilsimha) [inline online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala:26]: nit: configure your ide to use scalafmt.

**Review by nikhilsimha — COMMENTED** [review]
added minor comments. the general approach looks great! please go ahead and add unit tests.

File: online/src/main/scala/ai/chronon/online/FetcherBase.scala (hunk 1 above)
Comment (better365) [reply online/src/main/scala/ai/chronon/online/FetcherBase.scala:308]: we can change the log level to logger.debug

Comment (hanyuli1995) [reply online/src/main/scala/ai/chronon/online/FetcherBase.scala:308]: Yep these logs will be removed before merging, just use them for debugging

File: online/src/main/scala/ai/chronon/online/FetcherBase.scala [hunk 4]
```diff
@@ -288,29 +290,47 @@ class FetcherBase(kvStore: KVStore,
...
+                } catch {
+                  case ex: Exception =>
+                    // not all exceptions are due to stale schema, so we want to control how often we hit kv store
+                    getGroupByServingInfo.refresh(groupByServingInfo.groupByOps.metaData.name)
+                    context.incrementException(ex)
+                    ex.printStackTrace()
+                    throw ex
+                }
+              if (groupByServingInfo.groupBy.hasDerivations) {
+                val derivedMapTry: Try[Map[String, AnyRef]] = Try{
```
Comment (hanyuli1995) [inline online/src/main/scala/ai/chronon/online/FetcherBase.scala:316]: Trying to apply the new error handling logic we introduced in this PR: https://github.com/airbnb/chronon/pull/716

File: online/src/main/scala/ai/chronon/online/Fetcher.scala [hunk 5]
```diff
@@ -99,11 +100,18 @@ class Fetcher(val kvStore: KVStore,
                 val keyField = StructField(name, dType)
                 keyFields.add(keyField)
             }
-          val baseValueSchema = if (joinPart.groupBy.aggregations == null) {
+          val groupBySchemaBeforeDerivation: StructType = if (servingInfo.groupBy.aggregations == null) {
             servingInfo.selectedChrononSchema
           } else {
             servingInfo.outputChrononSchema
           }
+          val baseValueSchema: StructType = if (servingInfo.groupBy.derivations == null) {
```
Comment (better365) [inline online/src/main/scala/ai/chronon/online/Fetcher.scala:108]: Please use helper method [hasDerivations](https://github.com/airbnb/chronon/blob/main/api/src/main/scala/ai/chronon/api/Extensions.scala#L529)

File: online/src/main/scala/ai/chronon/online/FetcherBase.scala [hunk 6]
```diff
@@ -286,29 +288,48 @@ class FetcherBase(kvStore: KVStore,
...
+                val derivedMapTry: Try[Map[String, AnyRef]] = Try {
+                  val deriveFunc: DerivationFunc = getGroupByServingInfo(request.name).get.deriveFunc
+                  applyDeriveFunc(deriveFunc, request, groupByResponse)
+                }
+                val derivedMap = derivedMapTry match {
+                  case Success(derivedMap) =>
+                    derivedMap
+                  // If the derivation failed we want to return the exception map only
+                  case Failure(exception) => {
+                    val derivedExceptionMap =
```
Comment (better365) [inline online/src/main/scala/ai/chronon/online/FetcherBase.scala:323]: optional: do we want to also use the renaming logic here?

File: online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    val tsDsMap: Map[String, AnyRef] = {
+      Map("ts" -> (requestTs).asInstanceOf[AnyRef], "ds" -> (requestDs).asInstanceOf[AnyRef])
+    }
+    val derivedMap: Map[String, AnyRef] = Try(
+      deriveFunc(request.keys, baseMap ++ tsDsMap)
+        .mapValues(_.asInstanceOf[AnyRef])
+        .toMap) match {
+      case Success(derivedMap) => derivedMap
+      case Failure(exception) => {
+        throw exception
```
Comment (better365) [inline online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala:79]: let's not throw here. We could use datadog's increment exception method to capture it.

File: online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    if (derivationsScala.areDerivationsRenameOnly) {
+      val baseExpressions = if (derivationsScala.derivationsContainStar) {
+        baseValueSchema.filterNot { derivationsScala.derivationExpressionSet contains _.name }
+      } else {
+        Seq.empty
+      }
+      val expressions: Seq[StructField] = baseExpressions ++ derivationsScala.derivationsWithoutStar.map { d =>
+        {
+          if (baseValueSchema.typeOf(d.expression).isEmpty) {
+            throw new IllegalArgumentException(
```
Comment (better365) [inline online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala:114]: same as above. 
let's not throw here. We could use datadog's increment exception method to capture it.

File: online/src/main/scala/ai/chronon/online/Fetcher.scala (hunk 5 above)
Comment (hanyuli1995) [reply online/src/main/scala/ai/chronon/online/Fetcher.scala]: updated

File: online/src/main/scala/ai/chronon/online/FetcherBase.scala (hunk 6 above)
Comment (hanyuli1995) [reply online/src/main/scala/ai/chronon/online/FetcherBase.scala:327]: updated

File: online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala (hunk 8 above)
Comment (hanyuli1995) [reply online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala:108]: The exception here will be caught when we call the getJoinCodecs function, recorded in datadog, and then returned as join_codec_fetch_exception in the exception map https://sourcegraph.d.musta.ch/github.com/airbnb/chronon/-/blob/online/src/main/scala/ai/chronon/online/Fetcher.scala?L246

**Review by better365 — APPROVED** [review]
LGTM!

File: online/src/main/scala/ai/chronon/online/FetcherBase.scala [hunk 9]
```diff
@@ -292,29 +294,61 @@ class FetcherBase(kvStore: KVStore,
...
+              if (groupByServingInfo.groupBy.hasDerivations) {
+                val derivedMapTry: Try[Map[String, AnyRef]] = Try {
+                  applyDeriveFunc(groupByServingInfo.deriveFunc, request, groupByResponse)
+                }
+                val derivedMap = derivedMapTry match {
+                  case Success(derivedMap) =>
+                    derivedMap
+                  // If the derivation failed we want to return the exception map and rename only derivation
+                  case Failure(exception) => {
+                    val derivedExceptionMap =
```
Comment (hzding621) [inline online/src/main/scala/ai/chronon/online/FetcherBase.scala:328]: let's add exception metrics logging here

File: online/src/main/scala/ai/chronon/online/FetcherBase.scala [hunk 10]
```diff
@@ -292,29 +294,61 @@ class FetcherBase(kvStore: KVStore,
...
+                    val renameOnlyDerivedMapTry: Try[Map[String, AnyRef]] = Try {
+                      renameOnlyDeriveFunction(request.keys, groupByResponse)
+                        .mapValues(_.asInstanceOf[AnyRef])
+                        .toMap
+                    }
+                    // if the rename only derivation also failed we want to return the exception map
+                    val renameOnlyDerivedMap: Map[String, AnyRef] = renameOnlyDerivedMapTry match {
+                      case Success(renameOnlyDerivedMap) =>
+                        renameOnlyDerivedMap
+                      case Failure(exception) =>
```
Comment (hzding621) [inline online/src/main/scala/ai/chronon/online/FetcherBase.scala:341]: ditto, add exception metrics

**Review by hzding621 — APPROVED** [review]
LGTM w minor comments

**Pull request merged**